PSReadLine/Keys.cs

```diff
@@ -147,7 +157,8 @@ internal static void TryGetCharFromConsoleKey(ConsoleKeyInfo key, ref char resul
             {
                 flags |= (1 << 2); /* If bit 2 is set, keyboard state is not changed (Windows 10, version 1607 and newer) */
             }
-            int charCount = ToUnicode(virtualKey, scanCode, state, chars, chars.Length, flags);
+            uint layout = GetKeyboardLayout(GetWindowThreadProcessId(GetForegroundWindow(), 0));
```

Comment:
This always felt like a bug in console applications. I believe the conhost process receives the keyboard layout change notifications, but the `ToUnicode` call happens in a different process that never sees layout change notifications. I may have even opened a bug on `ToUnicode` that might have been resolved as won't fix over concerns about breaking something.

This change may seem work well enough in practice, but there is a race condition:

* Type a character
* New popup window gets created by some background process
* Call to `GetForegroundWindow` sees this new window instead of the console/vscode/etc. window.

I'm also not excited about seeing `GetWindowThreadProcessId` called on every key - it seems like it should be cached. Or maybe better - can we find the conhost/openconsole process instead? That process should see the correct layout, right?

Reply:
@lzybkr I create replacement for GetForegroundWindow.  What do you think about it?
```
// aka GetForegroundWindow
static Process findParentWithWindowHandle()
        {
            var current = Process.GetCurrentProcess();
            while (current != null)
            {
                if (current.MainWindowHandle != IntPtr.Zero)
                {
                    break;
                }
                current = GetParentProcess(current.Handle);
            }
            return current;
        }

 static Process GetParentProcess(IntPtr handle)
        {
            ProcessInformation processInformation = new ProcessInformation();

            int returnLength;
            int status = NtQueryInformationProcess(handle, 0, ref processInformation, Marshal.SizeOf(processInformation), out returnLength);
            if (status != 0)
                throw new Win32Exception(status);

            try
            {
                return Process.GetProcessById(processInformation.InheritedFromUniqueProcessId.ToInt32());
            }
            catch (ArgumentException)
            {
                // not found
                return null;
            }
        }
        [StructLayout(LayoutKind.Sequential)]
        public struct ProcessInformation
        {
            internal IntPtr Reserved1;
            internal IntPtr PebBaseAddress;
            internal IntPtr Reserved2_0;
            internal IntPtr Reserved2_1;
            internal IntPtr UniqueProcessId;
            internal IntPtr InheritedFromUniqueProcessId;

        }
```

Reply:
That looks like it would work well in Windows Terminal, VSCode, and the ISE, but it won't work with conhost because the parent will be explorer. conhost.exe is a sibling process.

Reply:
I wonder if you can use something like [GetProcessHandle](https://docs.microsoft.com/en-us/windows/win32/winauto/getprocesshandlefromhwnd)([GetConsoleWindow](https://docs.microsoft.com/en-us/windows/console/getconsolewindow)()) to get the correct process.

Reply:
I research this theme deeper and find out that all approaches that find window that subsribed to locale change does not work in all situations.

Reply:
I found that replacing https://github.com/PowerShell/PSReadLine/pull/3078/files#diff-9ab1ecb1e5d0ebd879a483eada4f6fcf2b407c662752f52bea29e45fd143f1a5R242
to c == '\0' do not break test.
And so far i didn't find dead key what start with different from '\0' code. Can it help?